tools/autograd/templates/ViewFuncs.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+// ${generated_comment}
+
+#include <torch/library.h>
+
+#include <c10/core/SymIntArrayRef.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Operators.h>
+#else
+$ops_headers
+#endif
+
+namespace torch { namespace autograd {
+
+// forward declaration so it can be a friend class of ViewFunc
+struct ViewFuncSavedStateGuard;
+
+/// Base class for view functions, providing reapplication of a view on a new base.
+/// Each view op should get a codegenerated subclass of this class containing
+/// any state needed to reconstruct the view. The class also provides convenience
+/// accessors for saved SymInts / tensor state. This is useful for e.g. fake-ification,
+/// where we want to use symbolic values or fake tensors instead.
+struct TORCH_API ViewFunc {
+  virtual ~ViewFunc() {}
+  /// Returns any SymInts in the saved state.
+  virtual std::vector<c10::SymInt> get_symints() const = 0;
+  /// Returns any Tensors in the saved state.
+  virtual std::vector<at::Tensor> get_tensors() const = 0;
+  /// Reapplies the view on the given base using the saved state.
+  /// If specified, the provided values for SymInts / Tensors
+  /// are hot-swapped in for the view call.
+  virtual at::Tensor operator()(
+      const at::Tensor&,
+      const std::optional<std::vector<c10::SymInt>>& = c10::nullopt,
+      const std::optional<std::vector<at::Tensor>>& = c10::nullopt) const = 0;
+  // Allow limited setter access to maintain the const-ness invariant.
+  friend class ViewFuncSavedStateGuard;
```

Comment:
The types on these functions are a bit suspicious and warrant further consideration.

Typically, you would want to return a non-owning reference for the getters (e.g., an ArrayRef) to avoid having to repeatedly copy vectors as is being done here. In contrast, the setter should take the input vector by value, so that a vector can be moved directly into the class without requiring a copy.

Reply:
Currently, the vectors are constructed on the fly for the `get_symints()` / `get_tensors()` calls. An alternative would be to initialize and always store the vectors in the ViewFunc derived classes; it would then be easy to return a reference type for the getters.

That said, I think there's some tradeoff here in terms of memory usage / computation when creating a ViewFunc vs. when calling it. I optimized for the former because I think ViewFuncs are created in a "hotter path" (including during eager) than when they are called (during fake-ification). wdyt?

Reply:
Yes, because fresh vectors can be allocated, return by value is correct here.